extensions/amp-base-carousel/0.1/child-layout-manager.js

```diff
@@ -0,0 +1,260 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const NO_INTERSECTION_MARGIN = '0%';
+
+/**
+ * The default margin around the scrolling element.
+ */
+const DEFAULT_NEARBY_MARGIN = '100%';
+
+/**
+ * What percentage of an Element must intersect before being considered as
+ * visible. This defaults to 1%, as the value for zero is not intuitive. For
+ * example, a horizontal scrolling container that has full width items will
+ * intersect on the second element if you have an intersection margin/threshold
+ * of 0, since the left edge of the second element "intersects" with the
+ * scrolling container, even though none of element is visible (not even 1 pixel).
+ *
+ * While the sepc says a value of "0" means "any non-zero number of pixels",
+ * there are actual no pixels actually visible at this threshold.
+ */
+const DEFAULT_INTERSECTION_THRESHOLD = 0.01;
+
+/**
+ * Manages scheduling layout/unlayout for children of an AMP component as they
+ * intersect with the AMP component. The parent AMP component should notify the
+ * manager as its own layout state changes so that the children can be updated
+ * accordingly.
+ * 
+ * Note: For Safari 12, this does not schedule layout for slides until they
+ * enter the viewport, since `rootMargin` on `IntersectionObserver` is not
+ * properly handled.
+ *
+ * Usage:
+ *
+ * Parent element:
+ * ```
+ * constructor() {
+ *   this.childLayoutManager = new ChildLayoutManager({
+ *     ampElement: this,
+ *   });
+ * }
+ 
+ * buildCallback() {
+ *   // Call this each time the effective children you want to manage change.
+ *   this.childLayoutManager.updateChildren(children);
+ * }
+ *
+ * layoutCallback() {
+ *   this.childLayoutManager.wasLaidOut();
+ * }
+ *
+ *
+ * unlayoutCallback() {
+ *   this.childLayoutManager.wasUnlaidOut();
+ * }
+ * ```
+ */
+export class ChildLayoutManager {
```

Comment:
Is this needed because Layers hasn't launched yet?

Reply:
Yep. When layers is ready, we might either take this away, or perhaps figure out if there is some value to having a carousel specific implementation (are there cases where we know what to do better than layers, which is more general?).

I think one (not very common, but possible) case is if you are say on the first slide and want to go to the nth slide (either by clicking a pagination dot, or some programmatic action). Since we smooth scroll, layers and this code will possibly (likely) load may slides along the way.

We could be a bit smarter and say to not load any slides along the way, and only load the destination slide.

Another possible difference is that we may not want to unlayout slides (or defer it) in case a user wants to go back and forth through a carousel. Currently, I believe runtime will unlayout things that exit the extended viewport area. This code does the same for now.

This actually differs from how carousel 0.1 works, which never does unlayout on slides (though it does `display: none` them).